tests.py

```diff
@@ -1,81 +1,241 @@
+        self.test_dummy_method = DummyMethod()
+        self.test_https = HTTPSMethod()
+
+        # Creation of Games
+        self.test_game_empty = Game(self.test_terminal_ui, self.test_empty)
+        self.test_game_https = Game(self.test_terminal_ui, self.test_https)
+        self.test_games = [self.test_game_https, self.test_game_empty]
+
+    def test_integration(self):
+        """ Checks if an empty victory works as expected """
+        for test_index, test_game in enumerate(self.test_games[1:]):
+            proc = subprocess.Popen(['python', '/Users/kbenarie/Projects/hangman-exercise/main.py'],
+                                    stdout=subprocess.PIPE)
+            stdout = proc.stdout.read().decode('utf-8')
+            assert stdout == EMPTY_SUCCESS, 'empty case did not succeed'
+
+    # Game Method Tests
+
+    def test_hangman_string_changes_when_needed_game_method(self):
+        letters = ['a', 5, 'b', 'd', 'ff', '']
+
+        for user_input in letters:
+            hangman_string_before_guess = self.test_https.initialize_method()
+            hangman_string_after_guess = self.test_https.play_one_turn(user_input).output
+            # checks if the edit distance is less or equal to 1
+            assert edit_distance(hangman_string_before_guess, hangman_string_after_guess) in {0, 1}
+
+    def test_guesses_number_goes_down_game_method(self):
+        letters = ['a', 5, 'b', 'd', 'ff', '']
+        guesses_number_before = self.test_https.play_one_turn(letters[0]).guesses_left
+
+        for user_input in letters[1:]:
+            guesses_number_after = self.test_https.play_one_turn(user_input).guesses_left
+            # checks if the guesses number allowed goes down
+            assert guesses_number_before >= guesses_number_after
+
+    def test_invalid_guess_game_method(self):
+        letters = ['4', 5, '', 'ff', '%']
+        response = self.test_https.play_one_turn(letters[0])
+        guesses_number_before, hangman_string_before_guess = response.guesses_left, response.output
+        for user_input in letters[1:]:
+            response = self.test_https.play_one_turn(user_input)
+            guesses_number_after, hangman_string_after_guess = response.guesses_left, response.output
+            assert guesses_number_after == guesses_number_before
+            assert hangman_string_after_guess == hangman_string_before_guess
+            assert response.current_mode == -1
+
+    def test_repeating_guess_game_method(self):
+        # repeating char - a
+        response = self.test_https.play_one_turn('a')
+        guesses_number_before, hangman_string_before_guess = response.guesses_left, response.output
+        response = self.test_https.play_one_turn('a')
+        guesses_number_after, hangman_string_after_guess = response.guesses_left, response.output
+        assert guesses_number_after == guesses_number_before
+        assert hangman_string_after_guess == hangman_string_before_guess
+
+    # UI Tests
```

Comment:
If these are UI tests, why do they use the fake UI? What exactly are we testing here?

Reply:
We spoke about this.